### ISSUE TITLE
4 point interval between my  finetune model and shared model

### ISSUE THREAD

**Issue by HuiBinR** [facebookresearch/GENRE#54]
I am doing the entity disambiguation task.
The model you published "[fairseq_entity_disambiguation_aidayago](**http://dl.fbaipublicfiles.com/GENRE/fairseq_entity_disambiguation_aidayago.tar.gz**)" get this result (without candidate and without the trie tree build specially for aida). This is almost same with  issue #26 but the aida-test-kilt result is a little different, #26 is 87.89, here is 87.92.

![1629877965(1)](https://user-images.githubusercontent.com/29992403/130750275-a2b83453-3511-4f03-bcec-51091fb7808e.jpg)

With the same code, my finetune model get this result (use the same WIKI trie tree 'kilt_titles_trie_dict.pkl'). For the test dateset aida-test-kilt, there is a ~4 point gap here.

![1629878164(1)](https://user-images.githubusercontent.com/29992403/130750525-15a79a98-08b9-4db3-b68e-b6fe2614cc74.jpg)

Although fairseq provide the default seed=1, each time i rerun the train.sh, the model is different, have different test result, but most of them are similar with the upon picture, where always have a ~4 point gap compare with the shared model.

Here is my finetune shell code. 
- Using the shared model 'fairseq_entity_disambiguation_blink' as pretrained model. 
- Based on the /GENRE/tree/main/scripts_genre/train.sh, i just **change the file path** and the **--max-update** 10000, **--total-num-update** 10000. So the finetune step is same as mentioned in the paper (10k). **
- Remove the --reset-meter** and **--reset-optimizer**, so the parameter is initialized from fairseq_entity_disambiguation_blink and the optimizer didn't changed.
- The code save 3 checkpoints: 50, 51, checkpoint_last. The  I test each of them and the best result is still around 83% for aida-test-kilt. The dict.source.txt and dict.target.txt is same with 'fairseq_entity_disambiguation_blink', which is just change the name of the 'dict.txt' file from the bart.large model. (I just copy the two file from 'fairseq_entity_disambiguation_blink' to the finetuned model)

````
DATASET=GENRE-main/datasets/aida
BASED_MODEL=fairseq_entity_disambiguation_blink
NAME=fairseq_aida_basedon_blink_10k_default
STEP=10000

fairseq-train $DATASET/bin/ \
    --save-dir GENRE-main/models/$NAME \
    --tensorboard-logdir tensorboard_logs/$NAME \
    --restore-file GENRE-main/models/$BASED_MODEL/model.pt \
    --arch bart_large  \
    --task translation  \
    --criterion label_smoothed_cross_entropy  \
    --source-lang source  \
    --target-lang target  \
    --truncate-source  \
    --label-smoothing 0.1  \
    --max-tokens 1024  \
    --update-freq 1  \
    --max-update $STEP  \
    --required-batch-size-multiple 1  \
    --dropout 0.1  \
    --attention-dropout 0.1  \
    --relu-dropout 0.0  \
    --weight-decay 0.01  \
    --optimizer adam  \
    --adam-betas "(0.9, 0.999)"  \
    --adam-eps 1e-08  \
    --clip-norm 0.1  \
    --lr-scheduler polynomial_decay  \
    --lr 3e-05  \
    --total-num-update $STEP  \
    --warmup-updates 500  \
    --ddp-backend no_c10d  \
    --num-workers 20  \ 
    --share-all-embeddings \
    --layernorm-embedding \
    --share-decoder-input-output-embed  \
    --skip-invalid-size-inputs-valid-test  \
    --log-format json  \
    --log-interval 10  \
    --patience 200  \
````

Here is my runing environment:  pytorch-1.6.0 ,  cuda10cudnn7 , python 3.7.7
I wanna know if something wrong with my finetune procedure? Since 83 to 87 has really a big gap.

Thanks for your work again~

**Comment by nicola-decao** [comment]
It looks like in the first table you are using a "trie tree build specially for aida" where in the second you are using "WIKI trie tree 'kilt_titles_trie_dict.pkl'". So these two tables do not look comparable.

**Comment by nicola-decao** [comment]
Try to lower the learning rate. I used many GPUs so the gradient might have been more precise.

**Comment by HuiBinR** [comment]
hi, i guess maybe this is because you use a smaller trie tree (tree build for aida) for pretrain and finetune, so the searching space is smaller for the model. This will bring some performance improvement.
Do you think this assumption is reasonable?

**Comment by nicola-decao** [comment]
I did not train with a trie. Not during pertaining nor during finetuning.

**Comment by HuiBinR** [comment]
Yes, you didn't. I re-read the section 4.1, you train and finetune just like normal generation task. But when testing, use a trie to constrain the output.
But in Appendix A.1, the paper mentioned `we fine-tune on AIDA without resetting the learning rate nor the optimizer statistics for 10k steps`.
I just use 1 GPU, and you use 32 GPU, will this lead the result difference? I will try to lower the learning rate and see.

**Comment by HuiBinR** [comment]
![1630396947(1)](https://user-images.githubusercontent.com/29992403/131465868-d1be88f1-245a-4b21-916a-2daa729b65e8.jpg)
As I read the `[fairseq document](https://fairseq.readthedocs.io/en/latest/command_line_tools.html#fairseq-train)`. They defaultly use distributed data parallel. You use 32 GPU, the batch size is 8 when `--required-batch-size-multiple 1` . So for each time of gradient descent, **the gradient is learned from 32*8=256 cases**. But when I use 1 gpu, the real batch size is 1*8=8. So this make some different.

I changed the learning rate, as it going lower, the performance getting better.  From the table below, we can say **1e-06** is a proper lr for 1 GPU.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/ruanhb2/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/ruanhb2/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">



  | wiki-test-kilt | msnbc-test-kilt | aida-test-kilt | clueweb-test-kilt | aquaint-test-kilt | ace2004-test-kilt
-- | -- | -- | -- | -- | -- | --
author_aida_model | 83.02 | 83.54 | 87.92 | 68.75 | 84.32 | 84.82
3e-05 | 76.53 | 78.66 | 83.86 | 66.59 | 81.84 | 81.71
2e-05 | 79.70 | 81.25 | 85.11 | 67.86 | 81.84 | 84.05
1e-05 | 82.01 | 81.40 | 85.77 | 68.74 | 84.32 | 84.44
7e-06 | 82.23 | 82.62 | 86.27 | 68.74 | 84.46 | 84.44
5e-06 | 82.44 | 82.62 | 86.24 | 68.92 | 84.32 | 84.44
3e-06 | 82.85 | 83.08 | 86.85 | 68.97 | 84.32 | 85.21
1e-06 | 82.82 | 83.69 | 87.54 | 69.06 | 85.14 | 85.99
9e-07 | 82.94 | 83.99 | 87.38 | 69.11 | 85.01 | 85.60



</body>

</html>





By the way. Thanks for you always quick answer for this issue and the issue #56 . I get the same trie tree with `kilt_titles_trie_dict.pkl` and get similar result with the model you shared.

Thanks again! Your guidances help a lot.
